internal/impl/deltalake/output.go

```diff
@@ -0,0 +1,202 @@
+		Fields(
+			service.NewStringField(dlFieldTableURL).
+				Description("The URL to table to ingest data into. Infers the storage backend to use from the scheme in the given table URL."),
+			service.NewOutputMaxInFlightField(),
+		)
+}
+
+func init() {
+	err := service.RegisterOutput("delta_lake", outputConfigSpec(),
+		func(conf *service.ParsedConfig, mgr *service.Resources) (out service.Output, maxInFlight int, err error) {
+			if maxInFlight, err = conf.FieldMaxInFlight(); err != nil {
+				return
+			}
+			out, err = newDeltaLakeWriterFromParsed(conf, mgr)
+			return
+		})
+	if err != nil {
+		panic(err)
+	}
+}
+
+func newDeltaLakeWriterFromParsed(conf *service.ParsedConfig, mgr *service.Resources) (*deltaLakeWriter, error) {
+	d := &deltaLakeWriter{logger: mgr.Logger()}
+
+	var err error
+	if d.tableURL, err = conf.FieldString(dlFieldTableURL); err != nil {
+		return nil, err
+	}
+
+	u, err := url.Parse(d.tableURL)
+	if err != nil {
+		return nil, err
+	}
+	d.bucket = u.Host
+	d.prefix = strings.TrimPrefix(u.Path, "/")
+
+	return d, nil
+}
+
+type deltaLakeWriter struct {
+	logger      *service.Logger
+	tableURL    string
+	bucket      string
+	prefix      string
+	table       *delta.Table
+	objectStore storage.ObjectStore
+	s3Client    *s3.Client
+}
+
+func (d *deltaLakeWriter) Connect(ctx context.Context) error {
+	cfg, err := config.LoadDefaultConfig(ctx)
+	if err != nil {
+		return fmt.Errorf("loading AWS config: %w", err)
+	}
+	s3Client := s3.NewFromConfig(cfg)
+
+	objectStorePath := storage.NewPath(d.tableURL)
+	objectStore, err := s3store.New(s3Client, objectStorePath)
+	if err != nil {
+		return fmt.Errorf("creating object store: %w", err)
+	}
+
+	stateStore := newLogStateStore(s3Client, d.bucket, d.prefix)
+	locker := nillock.New()
+	table, err := delta.OpenTable(objectStore, locker, stateStore)
+	if err != nil {
+		return fmt.Errorf("opening table: %w", err)
+	}
+
+	d.table = table
+	d.objectStore = objectStore
+	return nil
+}
+
+func (d *deltaLakeWriter) Write(ctx context.Context, msg *service.Message) error {
+	id := "2b0046bb-f9d1-46b1-9476-8eeaa6c27ce5"
+	// id := uuid.New().String()
+	fileName := fmt.Sprintf("part-%s.snappy.parquet", id)
+
+	msgBytes, err := msg.AsBytes()
+	if err != nil {
+		return fmt.Errorf("converting message to bytes: %w", err)
+	}
+
+	err = d.objectStore.Put(storage.NewPath(fileName), msgBytes)
+	if err != nil {
+		return fmt.Errorf("uploading object: %w", err)
+	}
+
+	add, _, err := delta.NewAdd(d.objectStore, storage.NewPath(fileName), make(map[string]string))
+	if err != nil {
+		return fmt.Errorf("creating new add operation: %w", err)
+	}
+
+	transaction := d.table.CreateTransaction(delta.NewTransactionOptions())
+	transaction.AddAction(add)
+	operation := delta.Write{Mode: delta.Append}
+	// TODO: metadata
+	transaction.SetOperation(operation)
+	version, err := transaction.Commit()
+	if err != nil {
+		return fmt.Errorf("committing transaction: %w", err)
+	}
+
+	fmt.Printf("version %d\n", version)
```

Comment:
Maybe we'd want to have debug log with more details?

Reply:
There is a follow up ticket to add metrics, I was going to flesh out logging in at that time